lib/ansible/plugins/connection/__init__.py

```diff
@@ -83,30 +83,30 @@ def __init__(self, play_context, new_stdin, shell=None, *args, **kwargs):
         self._connected = False
         self._socket_path = None
 
-        if shell is not None:
-            self._shell = shell
-
-        # load the shell plugin for this action/connection
-        if play_context.shell:
-            shell_type = play_context.shell
-        elif hasattr(self, '_shell_type'):
-            shell_type = getattr(self, '_shell_type')
-        else:
-            shell_type = 'sh'
-            shell_filename = os.path.basename(self._play_context.executable)
-            try:
-                shell = shell_loader.get(shell_filename)
-            except Exception:
-                shell = None
-            if shell is None:
-                for shell in shell_loader.all():
-                    if shell_filename in shell.COMPATIBLE_SHELLS:
-                        break
-            shell_type = shell.SHELL_FAMILY
-
-        self._shell = shell_loader.get(shell_type)
+        # helper plugins
+        self._shell = shell
+
+        # we always must have shell
         if not self._shell:
-            raise AnsibleError("Invalid shell type specified (%s), or the plugin for that shell type is missing." % shell_type)
+            self._shell = get_shell_plugin(shell_type=getattr(self, '_shell_type', None), executable=self._play_context.executable)
+
+        self._become = None
```

Comment:
Since we are using this outside of the connection (e.g. accessing `getattr(self._connection, '_become')`), we should remove the `_` prefix that indicates it is private.

Reply:
i would like to remedy that by moving most of those into connection/shell and out of action plugins, which should not care this much about these details.